### PR TITLE
feat: persist logos, animate tabs, add support for badges, fix minor display issues

### PR DESCRIPTION
**Pull request by daniel-mader**
# Description of change
Please write a summary of your changes and why you made them.

## Links to any relevant issues
Be sure to reference any related issues by adding `fixes issue #`.

## How the change has been tested
Describe the tests that you ran to verify your changes.
Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Definition of Done checklist
Add an `x` to the boxes that are relevant to your changes.

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src-tauri/src/state/reducers/dev_mode.rs [hunk 1]
```diff
@@ -12,19 +13,29 @@ use oid4vc_manager::methods::key_method::KeySubject;
...
+        })
+    );
```
Comment (nanderstabel) [inline src-tauri/src/state/reducers/dev_mode.rs:38]: Because this Badge Credential is added, both `test_load_dev_profile` and `test_load_dev_profile_twice` are failing. This can be simply solved by replacing the contents of `src-tauri/tests/fixtures/states/two_credentials_redirect_me.json` (and renaming the file to `src-tauri/tests/fixtures/states/three_credentials_redirect_me.json`) with:
```json
{
    "active_profile": {
        "name": "Ferris",
        "picture": "&#129408",
        "theme": "system",
        "primary_did": "did:key:z6Mkg1XXGUqfkhAKU1kVd1Pmw6UEj1vxiLj1xc91MBz5owNY"
    },
    "locale": "en",
    "credentials": [
        {
            "id": "33653133-6636-3633-6462-356231643533",
            "issuer_name": null,
            "format": {
                "format": "jwt_vc_json"
            },
            "data": {
                "@context": [
                    "https://www.w3.org/2018/credentials/v1",
                    "https://www.w3.org/2018/credentials/examples/v1"
                ],
                "type": [
                    "VerifiableCredential",
                    "DriverLicenseCredential"
                ],
                "issuer": "http://192.168.1.127:9090/",
                "issuanceDate": "2022-08-15T09:30:00Z",
                "expirationDate": "2027-08-15T23:59:59Z",
                "credentialSubject": {
                    "id": "did:key:z6Mkg1XXGUqfkhAKU1kVd1Pmw6UEj1vxiLj1xc91MBz5owNY",
                    "licenseClass": "Class C",
                    "issuedBy": "California",
                    "validity": "Valid"
                }
            }
        },
        {
            "id": "37626466-3738-6437-6165-623533613665",
            "issuer_name": null,
            "format": {
                "format": "jwt_vc_json"
            },
            "data": {
                "@context": [
                    "https://www.w3.org/2018/credentials/v1",
                    "https://www.w3.org/2018/credentials/examples/v1"
                ],
                "type": [
                    "VerifiableCredential",
                    "PersonalInformation"
                ],
                "issuanceDate": "2022-01-01T00:00:00Z",
                "issuer": "http://192.168.1.127:9090/",
                "credentialSubject": {
                    "id": "did:key:z6Mkg1XXGUqfkhAKU1kVd1Pmw6UEj1vxiLj1xc91MBz5owNY",
                    "givenName": "Ferris",
                    "familyName": "Crabman",
                    "email": "ferris.crabman@crabmail.com",
                    "birthdate": "1985-05-21"
                }
            }
        },
        {
            "id": "61363362-3264-3865-3331-653838616265",
            "issuer_name": null,
            "format": {
                "format": "jwt_vc_json"
            },
            "data": {
                "@context": [
                    "https://www.w3.org/2018/credentials/v1",
                    "https://www.w3.org/2018/credentials/examples/v1",
                    "https://purl.imsglobal.org/spec/ob/v3p0/context-3.0.2.json"
                ],
                "id": "http://example.edu/credentials/3732",
                "type": [
                    "VerifiableCredential",
                    "OpenBadgeCredential"
                ],
                "issuer": {
                    "id": "https://example.edu/issuers/565049",
                    "type": [
                        "IssuerProfile"
                    ],
                    "name": "Example University"
                },
                "issuanceDate": "2010-01-01T00:00:00Z",
                "name": "Teamwork Badge",
                "credentialSubject": {
                    "id": "did:example:ebfeb1f712ebc6f1c276e12ec21",
                    "type": [
                        "AchievementSubject"
                    ],
                    "achievement": {
                        "id": "https://example.com/achievements/21st-century-skills/teamwork",
                        "type": [
                            "Achievement"
                        ],
                        "criteria": {
                            "narrative": "Team members are nominated for this badge by their peers and recognized upon review by Example Corp management."
                        },
                        "description": "This badge recognizes the development of the capacity to collaborate within a group environment.",
                        "name": "Teamwork"
                    }
                }
            }
        }
    ],
    "current_user_prompt": {
        "type": "redirect",
        "target": "me"
    }
}
```

File: src-tauri/tests/tests/utils.rs
Comment (nanderstabel) [inline src-tauri/tests/tests/utils.rs:1]: These tests are all not compiling because of `fn download_asset`.

File: src-tauri/tests/tests/credential_offer.rs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+    assert!(send_credential_request(
+        state.borrow_mut(),
+        Action {
+            r#type: ActionType::CredentialOffersSelected,
+            payload: Some(json!({"offer_indices": [0]})),
+        },
+    )
+    .await
+    .is_ok());
+}
```
Comment (nanderstabel) [inline src-tauri/tests/tests/credential_offer.rs]: This test is failing because `send_credential_request` is expecting there to be a `CurrentUserPrompt::CredentialOffer` in the `AppState`. It uses the `offer_indices` it receives to find de credentials in the offer (possible misuses of `CurrentUserPrompt` but that's how it works.

This could be solved in the test by using this `AppState` instead:
```
    let mut state = AppState {
        managers: test_managers(vec![]),
        current_user_prompt: Some(CurrentUserPrompt::CredentialOffer {
            issuer_name: mock_server.uri().to_string(),
            logo_uri: Some("https://example.org/logo_uri".to_string()),
            credential_offer: CredentialOffer::<CredentialFormats> {
                credential_issuer: "https://example.org/credential_issuer".parse().unwrap(),
                credentials: vec![CredentialsObject::ByReference("UniversityDegreeCredential".to_string())],
                grants: None,
            },
        }),
        ..AppState::default()
    };
```

Obviously the `MockServer` needs to be able to respond to the respective calls that `send_credential_request` will make.

**Comment by nanderstabel** [comment]
`cargo clippy` gives a couple of warnings

File: src-tauri/src/state/reducers/credential_offer.rs [hunk 3]
```diff
@@ -52,8 +55,37 @@ pub async fn read_credential_offer(state: &mut AppState, action: Action) -> Resu
         None
     };
 
+    let credential_issuer_metadata = wallet
+        .get_credential_issuer_metadata(credential_issuer_url.clone())
+        .await
+        .ok();
```
Comment (nanderstabel) [inline src-tauri/src/state/reducers/credential_offer.rs:46]: Because of this code `test_qr_code_scanned_read_credential_offer` is failing. This is because that test relied on the fact that previously, in case the credential offer by value, there would not be a GET request to the issuer's metadata endpoint. With this addition the test will timeout because it cannot fetch any metadata (because it doesn't exists for that test). 
The code above your code:
```
    // Get the credential issuer metadata.
    let credential_issuer_metadata = if credential_offer
        .credentials
        .iter()
        .any(|credential| matches!(credential, CredentialsObject::ByReference(_)))
    {
        wallet
            .get_credential_issuer_metadata(credential_issuer_url.clone())
            .await
            .ok()
    } else {
        None
    };
```
ensured that there would be no HTTP request made by the `wallet` if not needed. 

However, since with your current update, we **do** need that metadata, regardless of whether the credential offer is by value or by reference. Therefore you can completely remove my code in lines 45 until 56. You can also completely remove `test_qr_code_scanned_read_credential_offer` since it is not 'supposed' to work anymore in it's current form.

So TL;DR:
* remove lines 45 until 56 in this file
* remove the `test_qr_code_scanned_read_credential_offer` test

File: src-tauri/src/state/reducers/credential_offer.rs [hunk 4]
```diff
@@ -258,6 +340,12 @@ pub async fn send_credential_request(state: &mut AppState, action: Action) -> Re
 
         info!("generated hash-key: {:?}", key);
 
+        persist_asset(
+            "6440ceac338a920197100e60_NGDIL%20Logo%20Dark.svg",
+            key.to_string().as_str(),
+        )
+        .ok();
```
Comment (nanderstabel) [inline src-tauri/src/state/reducers/credential_offer.rs]: What does this do? What is the significance of this: `"6440ceac338a920197100e60_NGDIL%20Logo%20Dark.svg"`?

File: src-tauri/src/state/reducers/credential_offer.rs [hunk 5]
```diff
@@ -94,13 +126,63 @@ pub async fn read_credential_offer(state: &mut AppState, action: Action) -> Resu
...
+    if credential_logo_url.is_some() {
+        debug!(
+            "{}",
+            format!(
+                "Downloading credential logo from url: {}",
+                credential_logo_url.unwrap().as_str().unwrap()
+            )
+        );
+        let _ = download_asset(credential_logo_url.unwrap().as_str().unwrap(), LogoType::CredentialLogo).await;
+    };
```
Comment (nanderstabel) [inline src-tauri/src/state/reducers/credential_offer.rs:169]: I'd go for:
```rust
    if let Some(logo_uri) = logo_uri.clone() {
        debug!("{}", format!("Downloading issuer logo from url: {}", logo_uri.as_str()));
        download_asset(logo_uri.as_str(), LogoType::IssuerLogo).await?;
    }
```

File: src-tauri/src/state/reducers/credential_offer.rs [hunk 6]
```diff
@@ -94,13 +126,63 @@ pub async fn read_credential_offer(state: &mut AppState, action: Action) -> Resu
...
+    if logo_uri.is_some() {
+        debug!(
+            "{}",
+            format!(
+                "Downloading issuer logo from url: {}",
+                logo_uri.clone().unwrap().as_str()
+            )
+        );
+        let _ = download_asset(logo_uri.as_ref().unwrap().as_str(), LogoType::IssuerLogo).await;
+    }
```
Comment (nanderstabel) [inline src-tauri/src/state/reducers/credential_offer.rs]: I'd go for:
```rust
    if let Some(logo_uri) = logo_uri.clone() {
        debug!("{}", format!("Downloading issuer logo from url: {}", logo_uri.as_str()));
        let _ = download_asset(logo_uri.as_str(), LogoType::IssuerLogo).await;
    }
```

File: src-tauri/src/state/reducers/credential_offer.rs (hunk 5 above)
Comment (nanderstabel) [inline src-tauri/src/state/reducers/credential_offer.rs]: I'd go for: 
```
    let credential_logo_url = first_credentials_supported_object.display.as_ref().and_then(|display| {
        display.first().and_then(|value| {
            value
                .get("logo")
                .and_then(|value| value.get("url").and_then(serde_json::Value::as_str))
        })
    });

    info!("credential_logo_url: {:?}", credential_logo_url);

    if let Some(credential_logo_url) = credential_logo_url {
        debug!(
            "{}",
            format!("Downloading credential logo from url: {}", credential_logo_url)
        );
        let _ = download_asset(credential_logo_url, LogoType::CredentialLogo).await;
    }
```

File: src-tauri/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+/// Downloads an asset to the system-specific data directory.
+/// The file is saved into the `assets/tmp` folder.
+/// Since the `/tmp` folder is cleared on each app restart,
+/// the downloaded file needs to be further treated and moved out of the `/tmp` folder.
+///
+/// Restrictions:
+/// - max. file size: 2MB
+/// - supported file types: `.png`, `.svg`
+pub async fn download_asset(url: &str, logo_type: LogoType) -> Result<(), AppError> {
+    let url: reqwest::Url = url.parse().unwrap();
```
Comment (nanderstabel) [inline src-tauri/src/utils.rs]: Probably cleaner to change this to:
```
pub async fn download_asset(url: reqwest::Url, logo_type: LogoType) -> Result<(), AppError> {
```

or alternatively replace the `unwrap()` for an error message:
```
    let url: reqwest::Url = url
        .parse()
        .map_err(|_| AppError::DownloadAborted("Could not parse URL string."))?;
```
edit: as long as we make sure that this function doesn't panic when it receives a string that is not a valid Url

File: src-tauri/tests/tests/credential_offer.rs [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+                                    },
+                                    "degree": {},
+                                    "gpa": {
+                                        "display": [
+                                            {
+                                                "name": "GPA"
+                                            }
+                                        ]
+                                    }
+                                })),
```
Comment (nanderstabel) [inline src-tauri/tests/tests/credential_offer.rs]: the occurrences of `credential_subject` in these test files are optional and not relevant for what's being tested, so I would suggest to set the values to `None` in order to make the tests a bit less bloated.

**Review by nanderstabel — APPROVED** [review]
Looks good!
